print-dsa.c

```diff
@@ -0,0 +1,235 @@
+ *
+ *      7   6   5   4   3   2   1   0  
+ *    .   .   .   .   .   .   .   .   .
+ *  0 +---+---+---+---+---+---+---+---+
+ *    |   Ether Destination Address   |
+ * +6 +---+---+---+---+---+---+---+---+
+ *    |     Ether Source Address      |
+ * +6 +---+---+---+---+---+---+---+---+  +-
+ *    |  Prog. DSA Ether Type [15:8]  |  | (8-byte) EDSA Tag
+ * +1 +---+---+---+---+---+---+---+---+  | Contains a programmable Ether type,
+ *    |  Prog. DSA Ether Type [7:0]   |  | two reserved bytes (always 0),
+ * +1 +---+---+---+---+---+---+---+---+  | and a standard DSA tag.
+ *    |     Reserved (0x00 0x00)      |  |
+ * +2 +---+---+---+---+---+---+---+---+  |  +-
+ *    | Mode  |b29|    Src/Trg Dev    |  |  | (4-byte) DSA Tag
+ * +1 +---+---+---+---+---+---+---+---+  |  | Contains a DSA tag mode,
+ *    |Src/Trg Port/Trunk |b18|b17|b16|  |  | source or target switch device,
+ * +1 +---+---+---+---+---+---+---+---+  |  | source or target port or trunk,
+ *    | PRI [2:0] |b12|  VID [11:8]   |  |  | and misc (IEEE and FPri) bits.
+ * +1 +---+---+---+---+---+---+---+---+  |  |
+ *    |           VID [7:0]           |  |  |
+ * +1 +---+---+---+---+---+---+---+---+  +- +-
+ *    |       Ether Length/Type       |
+ * +2 +---+---+---+---+---+---+---+---+
+ *    .   .   .   .   .   .   .   .   .
+ *
+ * Mode: Forward, To_CPU, From_CPU, To_Sniffer
+ * b29: (Source or Target) IEEE Tag Mode
+ * b18: Forward's Src_Is_Trunk, To_CPU's Code[2], To_Sniffer's Rx_Sniff
+ * b17: To_CPU's Code[1]
+ * b16: Original frame's CFI
+ * b12: To_CPU's Code[0]
+ */
+
+#define TOK(tag, byte, mask, shift) ((((u_char *) tag)[byte] & (mask)) >> (shift))
+
+#define DSA_LEN 4
+#define DSA_MODE(tag) TOK(tag, 0, 0xc0, 6)
+#define  DSA_MODE_TO_CPU 0x0
+#define  DSA_MODE_FROM_CPU 0x1
+#define  DSA_MODE_TO_SNIFFER 0x2
+#define  DSA_MODE_FORWARD 0x3
+#define DSA_TAGGED(tag) TOK(tag, 0, 0x20, 5)
+#define DSA_DEV(tag) TOK(tag, 0, 0x1f, 0)
+#define DSA_PORT(tag) TOK(tag, 1, 0xf8, 3)
+#define DSA_TRUNK(tag) TOK(tag, 1, 0x04, 2)
+#define DSA_RX_SNIFF(tag) TOK(tag, 1, 0x04, 2)
+#define DSA_CFI(tag) TOK(tag, 1, 0x01, 0)
+#define DSA_PRI(tag) TOK(tag, 2, 0xe0, 5)
+#define DSA_VID(tag) ((u_short)((TOK(tag, 2, 0xe0, 5) << 8) | (TOK(tag, 3, 0xff, 0))))
+#define DSA_CODE(tag) ((TOK(tag, 1, 0x06, 1) << 1) | TOK(tag, 2, 0x10, 4))
+
+#define EDSA_LEN 8
+#define EDSA_ETYPE(tag) ((u_short)((TOK(tag, 0, 0xff, 0) << 8) | (TOK(tag, 1, 0xff, 0))))
+
+static const struct tok dsa_mode_values[] = {
+	{ DSA_MODE_TO_CPU, "To CPU" },
+	{ DSA_MODE_FROM_CPU, "From CPU" },
+	{ DSA_MODE_TO_SNIFFER, "To Sniffer"},
+	{ DSA_MODE_FORWARD, "Forward" },
+	{ 0, NULL }
+};
+
+static const struct tok dsa_code_values[] = {
+	{ 0x0, "BPDU (MGMT) Trap" },
+	{ 0x1, "Frame2Reg" },
+	{ 0x2, "IGMP/MLD Trap" },
+	{ 0x3, "Policy Trap" },
+	{ 0x4, "ARP Mirror" },
+	{ 0x5, "Policy Mirror" },
+	{ 0, NULL }
+};
+
+static u_int
+dsa_if_print_full(netdissect_options *ndo, const struct pcap_pkthdr *h,
+		  const u_char *p, uint taglen)
```

Comment:
`uint` isn't guaranteed to be defined as a type, but `u_int` is; use `u_int`, so this builds on Windows.